### ISSUE TITLE
Tests stall consistently from different calls to `AwaitPromiseBuilder.wait`

### ISSUE THREAD

**Issue by ldiqual** [Quick/Nimble#499]
- [x] I have read [CONTRIBUTING](https://github.com/Quick/Nimble/blob/master/CONTRIBUTING.md) and have done my best to follow them.

### What did you do?

Running ~1700 tests that all spin up view controllers, assert using a mix of `toEventually/waitUntil`, fire some timers here and there, and make heavy use of `PromiseKit`. Plus a bunch of other third parties.

### What did you expect to happen?

Test suite to complete.

### What actually happened instead?

On iPhone 5s 10.3 simulator, most of the test runs stall (ie: xcode says a test is running but there's no progress for more than 30 seconds), every time in a different test. When pausing and firing the debugger, `AwaitPromiseBuilder.wait` seems to be the common "culprit" and most of the userland code triggers "toEventually" and "waitUntil".

Note that on iPhone 5s 11 simulator, tests complete more frequently if not always. 

I realize that those stalls can vary wildly depending on app code, timer usage, runloop manipulation etc. What I'm really looking for here is guidance on how to investigate this issue effectively. It seems that there are about ~25 timers on the runloop when I fire the debugger, but I have no idea if that's ok/too many/not relevant.

I guess the most confusing part to me is what Nimble really is doing in `wait` and how it's supposed to stop the test execution all while allowing the host app to keep running. I understand that it blocks the runloop from processing events, but I was under the impression that UIKit is using this runloop to do its thing so I don't understand how that would work. Is there a documentation – besides the one in Nimble's code – that would clarify that?

### Details

Result of pausing & running `RunLoop.current.getCFRunLoop()`: https://gist.github.com/ldiqual/fcb730b1cf0a2a3b4f0dae218cea7b8d

Stacktrace after pause: 
![image](https://user-images.githubusercontent.com/941453/36627810-66247c3a-18fd-11e8-8ed6-8b6f5e6e83b3.png)

App code in paused frame (one of many spots where it stalls):

![image](https://user-images.githubusercontent.com/941453/36627813-80efecfc-18fd-11e8-9335-ab2bf40f08b0.png)

Nimble code in paused frame: https://github.com/Quick/Nimble/blob/master/Sources/Nimble/Utils/Async.swift#L266

### Environment

 - Quick: *latest [master](https://github.com/Quick/Quick/commit/b172e8ad90fba4d21b4059297e047376939d5ca7)* because of https://github.com/Quick/Quick/pull/645
 - Nimble: *7.0.3*
 - Xcode Version: *9.2 (9C40b)*
 - Swift Version: *4.0*
 - Cocoapods: *1.3.1*

**Comment by AnthonyMDev** [comment]
I'm having very similar issues. It's happening relatively consistently, but not always in the same test. This occurs using `waitUntil` or `toEventually` for me. 
The tests stall out indefinitely. They don't ever timeout and fail or throw an exception or crash. It just sits there like it's a deadlock.

And the stacktrace always looks like this at the top:
```
#0	0x00007fff523b621a in mach_msg_trap ()
#1	0x00007fff523b67f8 in mach_msg ()
#2	0x00007fff23bd3cf5 in __CFRunLoopServiceMachPort ()
#3	0x00007fff23bcea17 in __CFRunLoopRun ()
#4	0x00007fff23bce066 in CFRunLoopRunSpecific ()
#5	0x00007fff2576b86f in -[NSRunLoop(NSRunLoop) runMode:beforeDate:] ()
#6	0x000000010f217cc5 in closure #3 in AwaitPromiseBuilder.wait(_:file:line:) at ~/Documents/repo/project/Pods/Nimble/Sources/Nimble/Utils/Await.swift:261
#7	0x000000010f20e96c in thunk for @callee_guaranteed () -> () ()
#8	0x000000010f20e9d1 in thunk for @escaping @callee_guaranteed () -> () ()
#9	0x000000010f208ff8 in -[NMBExceptionCapture tryBlock:] at ~/Documents/repos/project/Pods/Nimble/Sources/NimbleObjectiveC/NMBExceptionCapture.m:21
#10	0x000000010f217770 in AwaitPromiseBuilder.wait(_:file:line:) at ~/Documents/repos/project/Pods/Nimble/Sources/Nimble/Utils/Await.swift:251
#11	0x000000010f249879 in static NMBWait.throwableUntil(timeout:file:line:action:) at ~/Documents/repos/project/Pods/Nimble/Sources/Nimble/DSL+Wait.swift:67
#12	0x000000010f2492d1 in static NMBWait.until(timeout:file:line:action:) at ~/Documents/repos/project/Pods/Nimble/Sources/Nimble/DSL+Wait.swift:24
#13	0x000000010f24b6ac in waitUntil(timeout:file:line:action:) at ~/Documents/repos/project/Pods/Nimble/Sources/Nimble/DSL+Wait.swift:121
```

**Comment by AnthonyMDev** [comment]
I see this note in the README:

> Note: waitUntil triggers its timeout code on the main thread. Blocking the main thread will cause Nimble to stop the run loop to continue. This can cause test pollution for whatever incomplete code that was running on the main thread. Blocking the main thread can be caused by blocking IO, calls to sleep(), deadlocks, and synchronous IPC.

I'm not sure exactly what this means, or how to get around it. It seems like I am likely doing something to cause this, but the documentation here doesn't offer a suggestion on how to fix it.
I am using semaphores in my code on the main thread, but they don't cause a deadlock in my production code.

**Comment by ikesyo** [comment]
Marking this as a duplicate of #509 and fixed by #721.
